link/src/main/java/com/stripe/android/link/ui/signup/SignUpViewModel.kt

```diff
@@ -0,0 +1,195 @@
+                if (requiresNameCollection) {
+                    it.isValid()
+                } else {
+                    true
+                }
+            },
+            flow2 = emailController.fieldState.map { it.isValid() },
+            flow3 = phoneNumberController.isComplete
+        ) { nameComplete, emailComplete, phoneComplete ->
+            nameComplete && emailComplete && phoneComplete
+        }.collectLatest { formValid ->
+            updateState { it.copy(signUpEnabled = formValid) }
+        }
+    }
+
+    private suspend fun emailListener() {
+        emailController.formFieldValue.mapLatest { entry ->
+            entry.takeIf { it.isComplete }?.value
+        }.collectLatest { email ->
+            delay(LOOKUP_DEBOUNCE)
+            if (email != null) {
+                updateSignUpState(SignUpState.VerifyingEmail)
+                lookupConsumerEmail(email)
+            } else {
+                updateSignUpState(SignUpState.InputtingPrimaryField)
+            }
+        }
+    }
+
+    fun onSignUpClick() {
+        clearError()
+        viewModelScope.launch {
+            linkAccountManager.signUp(
+                email = emailController.fieldValue.value,
+                phone = phoneNumberController.fieldValue.value,
+                country = phoneNumberController.getCountryCode(),
+                name = nameController.fieldValue.value,
+                consentAction = SignUpConsentAction.Implied
+            ).fold(
+                onSuccess = {
+                    onAccountFetched(it)
+                    linkEventsReporter.onSignupCompleted()
+                },
+                onFailure = {
+                    onError(it)
+                    linkEventsReporter.onSignupFailure(error = it)
+                }
+            )
+        }
+    }
+
+    private fun onAccountFetched(linkAccount: LinkAccount?) {
+        if (linkAccount?.isVerified == true) {
+            navController?.popBackStack(LinkScreen.Wallet.route, inclusive = false)
+        } else {
+            navController?.navigate(LinkScreen.Verification.route)
+            // The sign up screen stays in the back stack.
+            // Clean up the state in case the user comes back.
+            emailController.onValueChange("")
+        }
+    }
+
+    private suspend fun lookupConsumerEmail(email: String) {
+        clearError()
+        linkAccountManager.lookupConsumer(email).fold(
+            onSuccess = {
+                if (it != null) {
+                    onAccountFetched(it)
+                } else {
+                    updateSignUpState(SignUpState.InputtingRemainingFields)
+                    linkEventsReporter.onSignupStarted()
+                }
+            },
+            onFailure = {
+                updateSignUpState(SignUpState.InputtingPrimaryField)
+                onError(it)
+            }
+        )
+    }
+
+    private fun onError(error: Throwable) {
+        logger.error("Error: ", error)
```

Comment:
should this logger indicate somehow that the error is coming from within the signup VM?